src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/PreferHtmlInAttributeValuesDocumentPositionStrategy.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
+using Microsoft.CodeAnalysis.Razor.Workspaces.Extensions;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.AutoInsert;
+
+// The main reason for this service is auto-insert of empty double quotes when a user types
+// equals "=" after Blazor component attribute. We think this is Razor (correctly I guess)
+// and wouldn't forward auto-insert request to HTML in this case. By essentially overriding
+// language info here we allow the request to be sent over to HTML where it will insert empty
+// double-quotes as it would for any other attribute value
+internal class PreferHtmlInAttributeValuesDocumentPositionInfoStrategy : IDocumentPositionInfoStrategy
+{
+    public static IDocumentPositionInfoStrategy Instance { get; } = new PreferHtmlInAttributeValuesDocumentPositionInfoStrategy();
+
+    public async Task<DocumentPositionInfo?> TryGetPositionInfoAsync(IRazorDocumentMappingService documentMappingService, DocumentContext documentContext, Position position, ILogger logger, CancellationToken cancellationToken)
+    {
+        var defaultDocumentPositionInfo = await DefaultDocumentPositionInfoStrategy.Instance.TryGetPositionInfoAsync(documentMappingService, documentContext, position, logger, cancellationToken).ConfigureAwait(false);
+        if (defaultDocumentPositionInfo is null)
+        {
+            return null;
+        }
+
+        if (defaultDocumentPositionInfo.LanguageKind != RazorLanguageKind.Razor ||
+            defaultDocumentPositionInfo.HostDocumentIndex < 1)
+        {
+            return defaultDocumentPositionInfo;
+        }
+
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+        var previousPosition = defaultDocumentPositionInfo.HostDocumentIndex - 1;
+        var charBeforePosition = codeDocument.GetSourceText()[previousPosition];
```

Comment:
❓ can this be done by checking the owner node instead of getting the full source text? I think full source text is less performant (or I may just have a bad mental model).

Reply:
I'm sure your mental model is much better than mine :) but looking at the implementation I'm not easily seeing an issue - it's just a wrapper around .Source.Text which are simple getters from what I can see

https://github.com/dotnet/razor/blob/0a163ded94e5c36a0f6c03279b9fff2b6658f9f4/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Extensions/RazorCodeDocumentExtensions.cs#L15

and Roslyn uses VS ITextImage for text storage and indexing, so I don't think there would be a large (or any) allocation to look up a single char. Please let me know if that's not the case, I'm sure you understand that code better than I do :) 

Can probably change the code to use a tree node lookup if needed.

Reply:
I doubt performance is an issue here, because the SourceText comes from LSP which sends us a giant string anyway, and to get a syntax tree, it would need the source text in order to parse anyway. If this was doing a ToString() on the SourceText then sure.

BUT I think we should use the syntax tree for this anyway. Strictly speaking this is IDE parsing the Razor document, and we should let the compiler do that :)

Reply:
Fair enough, added tree node and token usage there.